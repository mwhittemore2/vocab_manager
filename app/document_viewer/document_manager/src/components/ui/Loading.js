import React from "react"

/**
 * Renders a loading message for any component that
 * is experiencing high network latency.
 */
export class Loading extends React.Component {
    render(){
        return (
            <b>Loading...</b>
        )
    }
}