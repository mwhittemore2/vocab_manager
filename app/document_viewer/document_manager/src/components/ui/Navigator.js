import PropTypes from "prop-types"
import React from "react"
import C from '../../constants'

export const Navigator = ({cursor, jumpToPage, results, viewer}) =>
    <div id={viewer.concat("-navigator")}>
        <button id={viewer.concat("-previous")} onClick={cursor(C.PREVIOUS_PAGE, results.currPage)}>
            Previous
        </button>
        <div id={viewer.concat("-current")}>
           <input onKeyDown={(e) => jumpToPage(e)}
                  type="text"
                  placeholder={results.currPage}/>
        </div>
        <button id={viewer.concat("-next")} onClick={cursor(C.NEXT_PAGE, results.currPage)}>
            Next
        </button>
    </div>

Navigator.propTypes = {
    cursor: PropTypes.func,
    jumpToPage: PropTypes.func,
    results: PropTypes.object,
    viewer: PropTypes.string
}