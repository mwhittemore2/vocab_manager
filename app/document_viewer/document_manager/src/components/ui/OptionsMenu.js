import PropTypes from "prop-types"
import React from "react"
import stylesheet from './stylesheets/common'
import C from "../../constants"

const optionsMenuID = "document-viewer-options-menu"

export const DisplayOptions = ({selectOption}) =>
    <div id={optionsMenuID} className={stylesheet.option}>
        <b className={stylesheet.optionHeader}>Document Viewer Options</b>
        <br></br>
        <br></br>
        <ul className={stylesheet.chooseOption}>
            <div onClick={() => selectOption(C.DOC_VIEWER_OPTIONS.TRANSLATION_COORDINATOR)}>
                <li>Select words to translate</li>
            </div>
            <br></br>
            <div onClick={() => window.alert("This feature isn't implemented yet")}>
                <li>Manage bookmarks</li>
            </div>
        </ul>
    </div>

export class OptionsMenu extends React.Component{
    render(){
        let interaction = this.props.interaction
        if(interaction === C.DOCUMENT_VIEWER){
            let option = this.props.option
            if(option === C.DOC_VIEWER_OPTIONS.DEFAULT){
                let selectOption = this.props.selectOption
                return(
                    <DisplayOptions selectOption={selectOption}/>
                )
            }
        }
        return(
            <div id={optionsMenuID}></div>
        )
    }
}

DisplayOptions.propTypes = {
    selectOption: PropTypes.func
}

OptionsMenu.propTypes = {
    interaction: PropTypes.string,
    option: PropTypes.string,
    selectOption: PropTypes.func
}

/*
OptionsMenu.defaultProps = {
    interaction: C.DOCUMENT_VIEWER,
    option: C.DOC_VIEWER_OPTIONS.DEFAULT,
    selectOption: f=>f
}*/

export default OptionsMenu