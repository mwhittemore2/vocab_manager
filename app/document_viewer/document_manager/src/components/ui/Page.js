import React from "react"
import PropTypes from 'prop-types'
import { Word } from './Word'

const buildLineKey = (index, page) => {
    let domain = "line"
    let separator = "-"
    let key = domain + separator + page.currPage.toString() + separator + index.toString()
    return key
}

const buildWordKey = (index, word) => {
    let domain = "word"
    let separator = "-"
    let key = domain + separator + index.toString() + separator + word
    return key
}

const buildWord = (index, line, page, word) => {
    return {line: line, page: page.currPage, pos: index, text: word}
}

const canColor = (line, pos, page) => {
    let separator = ":"
    let word = page.currPage.toString() + separator + line.toString() + separator + pos.toString()
    let decision = page.selected.has(word)
    return decision
}

export const Line = ({line, num, page, selectWord}) => {
    return (
        <div>
            {line.map((word, index) => 
                <Word canColor={canColor(num, index, page)}
                      key={buildWordKey(index, word)}
                      selectWord={selectWord}
                      word={buildWord(index, num, page, word)}/>)}
            <br></br>
        </div>       
    )    
}

export const Page = ({lines={}, selectWord=f=>f}) =>
    <div className="page">
        {lines.words.map((line, index) =>
            <Line key ={buildLineKey(index, lines)}
                  line={line}
                  num={index}
                  page={lines}
                  selectWord={selectWord}/>)}
    </div>

Line.propTypes = {
    line: PropTypes.array,
    num: PropTypes.number,
    page: PropTypes.object,
    selectWord: PropTypes.func
}

Page.propTypes = {
    lines: PropTypes.object,
    selectWord: PropTypes.func
}