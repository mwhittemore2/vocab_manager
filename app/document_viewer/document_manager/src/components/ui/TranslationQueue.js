import PropTypes from "prop-types"
import React from "react"
import C from '../../constants'

const queueID = "translation-queue"

const buildKey = (index, word) => {
    let domain = "tq"
    let separator = "-"
    let key = domain + separator + index.toString() + separator + word
    return key
}

export const QueueElement = ({position, remove, text}) =>
    <span onClick={() => remove(position)}>{text}</span>

export class TranslationQueue extends React.Component{
    render(){
        let interaction = this.props.interaction
        if(interaction === C.DOCUMENT_VIEWER){
            let remove = this.props.remove
            let translations = this.props.translations
            return(
                <div id={queueID}>
                    {translations.searchPhrase.map((word, index) =>
                        <QueueElement key={buildKey(index, word.fulltext)}
                                      position={index}
                                      remove={remove}
                                      text={word.fulltext}/>)}
                </div>
            )
        }
        return(
            <div id={queueID}></div>
        )
    }
}

QueueElement.propTypes = {
    position: PropTypes.number,
    remove: PropTypes.func,
    text: PropTypes.string
}

TranslationQueue.propTypes = {
    interaction: PropTypes.string,
    remove: PropTypes.func,
    translations: PropTypes.object
}

export default TranslationQueue