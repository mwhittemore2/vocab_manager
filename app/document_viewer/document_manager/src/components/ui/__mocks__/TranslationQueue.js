import React from "react"

const TranslationQueueMock = () => <div></div>

TranslationQueueMock.displayName = "TranslationQueueMock"

export default TranslationQueueMock