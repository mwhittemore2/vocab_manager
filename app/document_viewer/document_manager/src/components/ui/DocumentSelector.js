import PropTypes from 'prop-types'
import React from "react"
import C from '../../constants'
import { Loading } from './Loading'

const buildKey = (doc, index) => {
    let domain = "docChoice"
    let separator = "-"
    let key = domain + separator + index.toString() + separator + doc
    return key
}

export const DocumentChoice = ({doc, selector}) =>
    <div onClick={() => selector(doc)}>
        {doc.title}
    </div>

export const DocumentIterator = ({documents, selectDocument}) =>
    <div id="document-selector">
        {documents.sort((a, b) => (a.title > b.title) ? 1 : -1).map((doc, index) =>
            <DocumentChoice key={buildKey(doc.title, index)}
                            doc={doc}
                            selector={selectDocument}/>
        )}
    </div>

 export class DocumentSelector extends React.Component {
    render(){
        let documents = this.props.documents
        let getDocuments = this.props.getDocuments
        let interaction = this.props.interaction
        let loaded = this.props.loaded[C.DOCUMENT_SELECTOR]
        let selectDocument = this.props.selectDocument
        if (interaction === C.DOCUMENT_SELECTOR){
            if (loaded){
                return (
                    <DocumentIterator documents={documents}
                                      selectDocument={selectDocument}/>
                )
            }
            else{
                if (documents.length <= 0){
                    getDocuments()
                }
                return (
                    <Loading />
                )
            }
        }
        return (
            <div id={"document-selector"}/>
        )
    }
}

DocumentChoice.propTypes = {
    doc: PropTypes.object,
    selector: PropTypes.func
}

DocumentIterator.propTypes = {
    documents: PropTypes.array,
    selectDocument: PropTypes.func
}

DocumentSelector.propTypes = {
    documents: PropTypes.array,
    getDocuments: PropTypes.func,
    interaction: PropTypes.string,
    loaded: PropTypes.object,
    selectDocument: PropTypes.func
}

DocumentSelector.defaultProps = {
    documents: [],
    getDocuments: f=>f,
    interaction: C.DOCUMENT_SELECTOR,
    loaded: {},
    selectDocument: f=>f
}

export default DocumentSelector