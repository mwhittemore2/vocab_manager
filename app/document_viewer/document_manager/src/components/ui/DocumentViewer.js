import PropTypes from "prop-types"
import React from "react"
import { Navigator } from './Navigator'
import { Page } from './Page'
import stylesheet from './stylesheets/common'
import C from '../../constants'

const displayID = "document-viewer"

/**
 * Manages the current page that the user is reading.
 * 
 * @param {func} cursor Function for moving to previous or next page.
 * @param {func} jumpToPage Function for moving to a specified page.
 * @param {object} lines The text of the currently displayed page.
 * @param {func} selectWord Adds the user-selected word to the translation 
 *                          queue.
 * @return {html} The HTML representation of the current page.
 */
export const Display = ({cursor, jumpToPage, lines, selectWord}) =>
    <div id={displayID} className={stylesheet.displayDocumentViewer}>
        <Page lines={lines} selectWord={selectWord} />
        <Navigator cursor={cursor}
                   jumpToPage={jumpToPage}
                   results={lines}
                   viewer={displayID}/>
    </div>

/**
 * Renders the interface for reading and traversing a document.
 */
export class DocumentViewer extends React.Component{
    render(){
        let interaction = this.props.interaction
        if(interaction === C.DOCUMENT_VIEWER){
            let cursor = this.props.cursor
            let jumpToPage = this.props.jumpToPage
            let lines = this.props.lines
            let selectWord = this.props.selectWord
            return(
                <Display cursor={cursor}
                         jumpToPage={jumpToPage}
                         lines={lines}
                         selectWord={selectWord}/>
            )
        }
        return(
            <div id={displayID}></div>
        )
    }
}

Display.propTypes = {
    cursor: PropTypes.func,
    jumpToPage: PropTypes.func,
    lines: PropTypes.object,
    selectWord: PropTypes.func
}

DocumentViewer.propTypes = {
    cursor: PropTypes.func,
    interaction: PropTypes.string,
    jumpToPage: PropTypes.func,
    lines: PropTypes.object,
    selectWord: PropTypes.func
}

DocumentViewer.defaultProps = {
    cursor: f=>f,
    interaction: C.DOCUMENT_SELECTOR,
    jumpToPage: f=>f,
    lines: {},
    selectWord: f=>f
}

export default DocumentViewer