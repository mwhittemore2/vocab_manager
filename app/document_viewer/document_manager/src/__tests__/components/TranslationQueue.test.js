import React from "react"
import { shallow } from "enzyme"
import toJSON from 'enzyme-to-json'
import { QueueElement, TranslationQueue } from '../../components/ui/TranslationQueue'
import C from '../../constants'

describe("QueueElement", ()=>{
    let position
    let remove
    beforeEach(()=> {
        position = 3
        remove = jest.fn()
    })

    afterAll(() => {
        jest.resetAllMocks()
    })

    it("single word", () => {
        let word = "Buch"
        let txt = toJSON(
                      shallow(<QueueElement position={position}
                                            remove={remove}
                                            text={word}/>
                      )
                  )
        expect(txt).toMatchSnapshot()
    })

    it("multiple words", () => {
        let words = "das Buch"
        let txt = toJSON(
                      shallow(<QueueElement position={position}
                                            remove={remove}
                                            text={words}/>
                      )
                  )
        expect(txt).toMatchSnapshot()
    })

    it("remove word", () =>{
        let word = "Buch"
        shallow(<QueueElement position={position}
                              remove={remove}
                              text={word}/>
        )
        .find('span')
        .simulate('click')
        expect(remove).toBeCalled()
    })
})

describe("TranslationQueue", () => {

    it("populated queue", () => {
        let interaction = C.DOCUMENT_VIEWER
        let translations = {
                searchPhrase: [
                    {fulltext: "Die "}, 
                    {fulltext: "Welt "}, 
                    {fulltext: "ist "}, 
                    {fulltext: "meine "}, 
                    {fulltext: "Vorstellung"}
                ]
            }
        let remove = jest.fn()
        let txt = toJSON(
                      shallow(<TranslationQueue interaction={interaction}
                                                remove={remove}
                                                translations={translations}/>
                      )
                  )
        expect(txt).toMatchSnapshot()
    })

    it("empty queue", () => {
        let interaction = C.DOCUMENT_VIEWER
        let translations = {
            searchPhrase: []
        }
        let remove = jest.fn()
        let txt = toJSON(
                      shallow(<TranslationQueue interaction={interaction}
                                                remove={remove}
                                                translations={translations}/>
                      )
                  )
        expect(txt).toMatchSnapshot()
    })

    it("not in document viewer", () => {
        let interaction = C.DOCUMENT_SELECTOR
        let translations = {
            searchPhrase: []
        }
        let remove = jest.fn()
        let txt = toJSON(
                      shallow(<TranslationQueue interaction={interaction}
                                                remove={remove}
                                                translations={translations}/>
                      )
                  )
        expect(txt).toMatchSnapshot()
    })
})