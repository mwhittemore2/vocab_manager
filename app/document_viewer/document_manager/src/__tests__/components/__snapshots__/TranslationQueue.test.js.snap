// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`QueueElement multiple words 1`] = `
<span
  onClick={[Function]}
>
  das Buch
</span>
`;

exports[`QueueElement single word 1`] = `
<span
  onClick={[Function]}
>
  Buch
</span>
`;

exports[`TranslationQueue empty queue 1`] = `
<div
  className="display-queue"
  id="translation-queue"
/>
`;

exports[`TranslationQueue not in document viewer 1`] = `
<div
  id="translation-queue"
/>
`;

exports[`TranslationQueue populated queue 1`] = `
<div
  className="display-queue"
  id="translation-queue"
>
  <QueueElement
    key="tq-0-Die "
    position={0}
    remove={[MockFunction]}
    text="Die "
  />
  <QueueElement
    key="tq-1-Welt "
    position={1}
    remove={[MockFunction]}
    text="Welt "
  />
  <QueueElement
    key="tq-2-ist "
    position={2}
    remove={[MockFunction]}
    text="ist "
  />
  <QueueElement
    key="tq-3-meine "
    position={3}
    remove={[MockFunction]}
    text="meine "
  />
  <QueueElement
    key="tq-4-Vorstellung"
    position={4}
    remove={[MockFunction]}
    text="Vorstellung"
  />
</div>
`;
